File: java/src/main/java/com/example/doubleup/service/PlayService.java
package com.example.doubleup.service;

import com.example.doubleup.dao.GameRepository;
import com.example.doubleup.dao.PlayerRepository;
import com.example.doubleup.dto.PlayResponseDto;
import com.example.doubleup.enums.AuditOperation;
import com.example.doubleup.enums.GameResult;
import com.example.doubleup.enums.SmallLargeChoice;
import com.example.doubleup.model.Game;
import com.example.doubleup.model.Player;
import jakarta.transaction.Transactional;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.security.SecureRandom;
import java.time.OffsetDateTime;
import java.util.Optional;
import java.util.UUID;

@Service
public class PlayService {

    @Autowired
    private PlayerRepository playerRepository;

    @Autowired
    private GameRepository gameRepository;

    @Autowired
    private AuditLogService auditLogService;

    public Player getPlayer(UUID id) throws Exception {
        Optional<Player> playerOptional = playerRepository.findById(id);
        if (playerOptional.isEmpty()) {
            throw new Exception("Player id: " + id + " not found!");
        }
        Player player = playerOptional.get();
        auditLogService.writeAuditLog(player, AuditOperation.SELECT, player.getId(), "player");
        return player;
    }

    private PlayResponseDto processGame(Long betSize, SmallLargeChoice playerChoice, Long accountBalance) {
        SecureRandom secureRandom = new SecureRandom();
        int randomNumber = secureRandom.nextInt(13) + 1;

        // generate a random number to ensure cryptographic-strength randomness
        SmallLargeChoice gameChoice = null;
        if (randomNumber <= 6) {
            gameChoice = SmallLargeChoice.SMALL;
        } else if (randomNumber >= 8) {
            gameChoice = SmallLargeChoice.LARGE;
        }

        // decide the game result
        GameResult gameResult = GameResult.L;
        if (gameChoice == playerChoice) {
            gameResult = GameResult.W;
        }

        // money in play
        Long moneyInPlay = gameResult == GameResult.W ? betSize * 2 : 0;

        return new PlayResponseDto((short)randomNumber, gameResult, moneyInPlay, accountBalance);
    }

    public PlayResponseDto playGame(Player player, Long betSize, SmallLargeChoice choice) {
        PlayResponseDto playResponseDto = processGame(betSize, choice, player.getAccountBalance());
        player.setMoneyInPlay(playResponseDto.getMoneyInPlay());
        playerRepository.save(player);
        auditLogService.writeAuditLog(player, AuditOperation.UPDATE, player.getId(), "player");

        Game game = new Game(player, OffsetDateTime.now(), betSize, choice, playResponseDto.getCardDrawn(), betSize * 2, playResponseDto.getGameResult());
        gameRepository.save(game);
        auditLogService.writeAuditLog(player, AuditOperation.INSERT, game.getId(), "game");

        return playResponseDto;
    }

    @Transactional
    public PlayResponseDto startGame(Player player, Long betSize, SmallLargeChoice choice) {
        player.setAccountBalance(player.getAccountBalance() - betSize);
        return playGame(player, betSize, choice);
    }

    @Transactional
    public PlayResponseDto continueGame(Player player, Long betSize, SmallLargeChoice choice) {
        return playGame(player, betSize, choice);
    }
}

File: go/service/play_service.go
package service

import (
    "fmt"
    "database/sql"
    "github.com/google/uuid"
    "math/rand"
    "time"
    "double_up/dao"
    "double_up/dto"
    "double_up/enums"
    "double_up/model"
)

func GetPlayer(db *sql.DB, id uuid.UUID) (*model.Player, error) {
    player, err := dao.FindPlayerById(db, id)
    if err != nil {
        fmt.Println(err)
        return nil, err
    }
    WriteAuditLog(db, player, enums.Select, player.ID, "player")
    return player, err
}

func ProcessGame(randomNumber int, betSize int64, playerChoice enums.SmallLargeChoice, accountBalance int64) *dto.PlayResponseDto {

	// decide the small / large choice based on number generated
	var gameChoice enums.SmallLargeChoice
	if randomNumber <= 6 {
		gameChoice = enums.Small
	} else if randomNumber >= 8 {
		gameChoice = enums.Large
	}

    // decide the game result
	gameResult := enums.L
	if gameChoice == playerChoice {
		gameResult = enums.W
	}

    // money in play
	var moneyInPlay int64
	if gameResult == enums.W {
		moneyInPlay = betSize * 2
	} else {
		moneyInPlay = 0
	}

	return &dto.PlayResponseDto{
		CardDrawn:      int16(randomNumber),
		GameResult:     gameResult,
		MoneyInPlay:    moneyInPlay,
		RemainingBalance: accountBalance,
	}
}

func GenerateNumberAndProcessGame(betSize int64, playerChoice enums.SmallLargeChoice, accountBalance int64) *dto.PlayResponseDto {
    // generate a random number to ensure cryptographic-strength randomness
    rand.Seed(time.Now().UnixNano())
    randomNumber := rand.Intn(13) + 1
	return ProcessGame(randomNumber, betSize, playerChoice, accountBalance)
}

func PlayGame(db *sql.DB, player *model.Player, betSize int64, choice enums.SmallLargeChoice) (*dto.PlayResponseDto, error) {
    playResponseDto := GenerateNumberAndProcessGame(betSize, choice, player.AccountBalance)
    player.MoneyInPlay = playResponseDto.MoneyInPlay
    dao.UpdatePlayer(db, player)
    WriteAuditLog(db, player, enums.Update, player.ID, "player")

    game := model.NewGame(player.ID, betSize, choice, playResponseDto.CardDrawn, betSize * 2, playResponseDto.GameResult)
    dao.CreateGame(db, game)
    WriteAuditLog(db, player, enums.Insert, game.ID, "game")

    return playResponseDto, nil
}

func StartGame(db *sql.DB, player *model.Player, betSize int64, choice enums.SmallLargeChoice) (*dto.PlayResponseDto, error) {
    player.AccountBalance -= betSize
    return PlayGame(db, player, betSize, choice)
}

func ContinueGame(db *sql.DB, player *model.Player, betSize int64, choice enums.SmallLargeChoice) (*dto.PlayResponseDto, error) {
    return PlayGame(db, player, betSize, choice)
}
